qiskit/chemistry/ground_state_calculation/adapt_vqe.py

```diff
@@ -0,0 +1,271 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+A ground state calculation employing the AdaptVQE algorithm.
+"""
+
+from typing import Optional, List, Tuple
+import logging
+import re
+import numpy as np
+
+from qiskit.aqua import AquaError
+from qiskit.aqua.algorithms import VQE
+from qiskit.aqua.operators import LegacyBaseOperator, WeightedPauliOperator
+from qiskit.aqua.utils.validation import validate_min
+from qiskit.chemistry.components.variational_forms import UCCSD
+from qiskit.chemistry.drivers import BaseDriver
+from qiskit.chemistry.qubit_transformations import FermionicTransformation
+from qiskit.chemistry.results import ElectronicStructureResult
+
+from .ground_state_calculation import GroundStateCalculation
+from .mes_factories import VQEUCCSDFactory
+
+logger = logging.getLogger(__name__)
+
+
+class AdaptVQE(GroundStateCalculation):
+    """A ground state calculation employing the AdaptVQE algorithm."""
+
+    def __init__(self,
+                 transformation: FermionicTransformation,
+                 solver: VQEUCCSDFactory,
```

Comment:
Yes. The reason for fixing it is that AdaptVQE only works with VQE and UCCSD. If you take a look at the unittests you can see an example of how one can simply change things like the optimizer.

Reply:
I looked at the unit test - replacing the method of the class by some other method - we cannot expect users to do this! I would expect a user to create a whole new MESFactory subclass - while VQEUCCSDFactory could be overridden it would just be for its type and you check the solver later. So I would make this an MESFactory here if we have to have some factory.

Reply:
Okay I agree that we shouldn't enforce this type on the user. I changed it to simply except an `MESFactory`.